api/clients/eigenda_client_v2.go

```diff
@@ -0,0 +1,142 @@
+	if clientConfig.DisablePointVerificationMode {
+		codec = codecs.NewNoIFFTCodec(lowLevelCodec)
+	} else {
+		codec = codecs.NewIFFTCodec(lowLevelCodec)
+	}
+
+	return NewEigenDAClientV2(log, clientConfig, relayClient, codec)
+}
+
+// NewEigenDAClientV2 assembles an EigenDAClientV2 from subcomponents that have already been constructed and initialized.
+func NewEigenDAClientV2(
+	log logging.Logger,
+	clientConfig EigenDAClientConfig,
+	relayClient RelayClient,
+	codec codecs.BlobCodec) (EigenDAClientV2, error) {
+
+	return &eigenDAClientV2{
+		clientConfig: clientConfig,
+		log:          log,
+		relayClient:  relayClient,
+		codec:        codec,
+	}, nil
+}
+
+// GetBlob iteratively attempts to retrieve a given blob with key blobKey from the relays listed in the blobCertificate.
+//
+// The relays are attempted in random order.
+//
+// The returned blob is decoded.
+func (c *eigenDAClientV2) GetBlob(ctx context.Context, blobKey corev2.BlobKey, blobCertificate corev2.BlobCertificate) ([]byte, error) {
+	// create a randomized array of indices, so that it isn't always the first relay in the list which gets hit
+	random := rand.New(rand.NewSource(rand.Int63()))
+	relayKeyCount := len(blobCertificate.RelayKeys)
+	var indices []int
+	for i := 0; i < relayKeyCount; i++ {
+		indices = append(indices, i)
+	}
+	random.Shuffle(len(indices), func(i int, j int) {
+		indices[i], indices[j] = indices[j], indices[i]
+	})
+
+	// TODO (litt3): consider creating a utility which can deprioritize relays that fail to respond (or respond maliciously)
+
+	// iterate over relays in random order, until we are able to get the blob from someone
+	for _, val := range indices {
+		relayKey := blobCertificate.RelayKeys[val]
+
+		data, err := c.relayClient.GetBlob(ctx, relayKey, blobKey)
+
+		// if GetBlob returned an error, try calling a different relay
+		if err != nil {
+			// TODO: should this log type be downgraded to debug to avoid log spam? I'm not sure how frequent retrieval
+			//  from a relay will fail in practice (?)
+			c.log.Info("blob couldn't be retrieved from relay", "blobKey", blobKey, "relayKey", relayKey)
+			continue
+		}
+
+		// An honest relay should never send an empty blob
+		if len(data) == 0 {
+			c.log.Warn("blob received from relay had length 0", "blobKey", blobKey, "relayKey", relayKey)
+			continue
+		}
+
+		// An honest relay should never send a blob which cannot be decoded
+		decodedData, err := c.codec.DecodeBlob(data)
+		if err != nil {
+			c.log.Warn("error decoding blob", "blobKey", blobKey, "relayKey", relayKey)
+			continue
+		}
+
+		return decodedData, nil
+	}
+
+	return nil, fmt.Errorf("unable to retrieve blob from any relay")
+}
+
+func (c *eigenDAClientV2) GetCodec() codecs.BlobCodec {
+	return c.codec
+}
+
+func (c *eigenDAClientV2) Close() error {
+	var errList *multierror.Error
+
+	// TODO: this is using a multierror, since there will be more subcomponents requiring closing after adding PUT functionality
```

Comment:
multierror is very old. the stdlib's error package can now do this, see https://pkg.go.dev/errors#Join

Reply:
Switched to use join